### PR TITLE
Start adding type annotations to `astropy/units/core.py`

### PR DESCRIPTION
**Pull request by eerovaher**
### Description

I have added some type annotations to `astropy/units/core.py`. I only added those annotations I thought were simple to add, so the fraction of annotated code is still quite low, but this pull request is large enough as it is. Furthermore, type checkers become more useful as the fraction of annotated code increases, so once the simple annotations are written down it should become easier to add the more complex ones.

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/units/core.py [hunk 1]
```diff
@@ -787,7 +801,7 @@ def _normalize_equivalencies(equivalencies):
 
         return normalized
 
-    def __pow__(self, p):
+    def __pow__(self, p: Real) -> CompositeUnit:
```
Comment (eerovaher) [inline astropy/units/core.py:804]: This is a case where the annotation is stricter then what the code actually allows at runtime, but I am not convinced that array powers should be supported at all and if anyone does get spooked into changing a constant-valued array to a scalar in their code because of this annotation then I don't think that's a bad thing at all.

**Review by mhvk — APPROVED** [review]
Nice!  I've gone one comment only, which can be ignored if needed, so I'll approve. But @nstarman definitely should have a good look too.

File: astropy/units/core.py [hunk 2]
```diff
@@ -687,27 +697,31 @@ def aliases(self):
...
+    def powers(self) -> list[UnitPower]:
         """
         Return the powers of the unit.
         """
         return [1]
 
-    def to_string(self, format=unit_format.Generic, **kwargs):
+    def to_string(
+        self,
+        format: type[unit_format.Base] | str | None = unit_format.Generic,
```
[the remark this answers is not in the document]
Comment (eerovaher) [reply astropy/units/core.py:722]: There are two good reasons for allowing `None` here.

1. There is code in `astropy` that can call `UnitBase.to_string()` with `format=None`, e.g. https://github.com/astropy/astropy/blob/6130037fc72111ec24c31421884b7859f2b02745/astropy/coordinates/angles/core.py#L356

2. Calling `UnitBase.to_string()` with `format=None` leads to fastest execution at runtime because this function calls `astropy.units.format.get_format()` and the first thing that function does is it checks https://github.com/astropy/astropy/blob/6130037fc72111ec24c31421884b7859f2b02745/astropy/units/format/__init__.py#L90 

Both of these points I brought up could be changed, but the annotation correctly documents the current situation.

[the remark this answers is not in the document]
Comment (eerovaher) [reply astropy/units/core.py:722]: There are many dunder methods in this file with docstrings, even though dunder methods don't really need them because what they do is defined well enough by the [Python data model](https://docs.python.org/3/reference/datamodel.html) (`__new__()`, `__init__()` and `__call__()` are the three exceptions to this).

Additionally, there are many properties with docstrings like https://github.com/astropy/astropy/blob/6130037fc72111ec24c31421884b7859f2b02745/astropy/units/core.py#L680-L684
when it would be much better to have a docstring like
```python
     @property 
     def aliases(self): 
         """The alias (long) names for this unit.""" 
```

In short, there are many changes to be made to the docstrings here, not just the change you are suggesting.

My preference would be to limit `typing` pull requests only to annotations and to address whatever the annotations reveal in follow-up pull requests, but if you or @nstarman think differently then let me know and I can push more commits here.

[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/units/core.py:722]: > My preference would be to limit `typing` pull requests only to annotations and to address whatever the annotations reveal in follow-up pull requests, but if you or @nstarman think differently then let me know and I can push more commits here.

As long as we don't forget, I'm happy either way! (And at my age, forgetting is likely enough that I typically try to fix things as I go...)

As said, though, if @nstarman is happy, I think we should just merge this rather than run CI another time just for a really minor docstring fix.

[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/units/core.py:722]: Agreed!

**Review by nstarman — APPROVED** [review]
🚀

**Comment by mhvk** [comment]
OK, let's get it in. Thanks, @eerovaher!
